docs/frontend-performance.md

```diff
@@ -0,0 +1,2984 @@
+  - [❌ Anti-Pattern: Large Object Retention in Closures](#-anti-pattern-large-object-retention-in-closures)
+
+### Chapter 3: State Management
+
+- [Advanced Selector Patterns](#advanced-selector-patterns)
+  - [❌ Anti-Pattern: Identity Functions as Output Selectors](#-anti-pattern-identity-functions-as-output-selectors)
+  - [❌ Anti-Pattern: Selectors That Return Entire State Objects](#-anti-pattern-selectors-that-return-entire-state-objects)
+  - [❌ Anti-Pattern: Selectors Without Proper Input Selectors](#-anti-pattern-selectors-without-proper-input-selectors)
+  - [Best Practices for Reselect Selectors](#best-practices-for-reselect-selectors)
+  - [Use `createDeepEqualSelector` sparingly](#use-createdeepequalselector-sparingly)
+  - [Combine related selectors into one memoized selector](#combine-related-selectors-into-one-memoized-selector)
+  - [❌ Anti-Pattern: Inline Selector Functions in useSelector](#-anti-pattern-inline-selector-functions-in-useselector)
+  - [❌ Anti-Pattern: Multiple useSelector Calls Selecting Same State Slice](#-anti-pattern-multiple-useselector-calls-selecting-same-state-slice)
+  - [❌ Anti-Pattern: Inefficient Use of `Object.values()` and `Object.keys()` in Selectors](#-anti-pattern-inefficient-use-of-objectvalues-and-objectkeys-in-selectors)
+  - [❌ Anti-Pattern: Deep Property Access in Selectors](#-anti-pattern-deep-property-access-in-selectors)
+  - [❌ Anti-Pattern: Repeated Object Traversal in Selectors](#-anti-pattern-repeated-object-traversal-in-selectors)
+  - [❌ Anti-Pattern: Selectors That Reorganize Nested State](#-anti-pattern-selectors-that-reorganize-nested-state)
+  - [❌ Anti-Pattern: Filtering/Searching Through Nested Objects](#-anti-pattern-filteringsearching-through-nested-objects)
+
+### Chapter 4: React Compiler
+
+- [React Compiler Considerations](#react-compiler-considerations)
+  - [What React Compiler Does](#what-react-compiler-does)
+  - [React Compiler Assumptions](#react-compiler-assumptions)
+  - [React Compiler Limitations](#react-compiler-limitations)
+  - [When Manual Memoization is Still Required](#when-manual-memoization-is-still-required)
+  - [Decision Tree: Do You Need Manual Memoization?](#decision-tree-do-you-need-manual-memoization)
+  - [Summary: React Compiler Capabilities and Limitations](#summary-react-compiler-capabilities-and-limitations)
+
+---
+
+## Optimizing Lists and Large Datasets
+
+### Use Proper Keys
+
+```typescript
+❌ WRONG: Using array index as key
+const TokenList = ({ tokens }: TokenListProps) => {
+  return (
+    <div>
+      {tokens.map((token, index) => (
+        <TokenItem key={index} token={token} /> // Bad if list can reorder!
+      ))}
+    </div>
+  );
+};
+
+✅ CORRECT: Use unique, stable identifiers
+const TokenList = ({ tokens }: TokenListProps) => {
+  return (
+    <div>
+      {tokens.map(token => (
+        <TokenItem key={token.address} token={token} />
```

Comment:
Are there any advices for the following scenarios:
- What if we don't have unique identifiers?
- How effectively these patterns apply to nested lists and what would be advice for keys in such cases?
- Are constructions similar to this `key={`${selectedAlert.key}-detail-${index}`}` good or bad practice and in which cases these can be possibly good or bad?

Do you think that maybe some of these scenarios are worth mentioning? They do happen across the project and are part of our codebase at present.